### DISCUSSION TITLE
Version 3.3.0 is now available.

### DISCUSSION THREAD

**Discussion by janiversen** [pymodbus-dev/pymodbus#1570] [Announcements]
This version contains 73 commits made by 9 committers.

The version contains, apart from bug fixes, 3 bigger items:
- The async clients have a new transport layer, with common reconnect, timeout etc.
- Precommit and ruff have been added to the CI as well as the local "git commit".
- Type hinting have been added to several modules
Please make sure to have a look at API_CHANGES.rst to see the changes affecting applications.

Thanks a lot to all that helped make this release possible, especially all the new committers.

Version 3.4 will focus on:
- A rewrite of the simulator configuration and additional documentation with more examples
- A common message layer for the async clients
the target for the release is July / August

**Comment by alexrudd2** [comment]
> Type hinting have been added to several modules

This is great, the new type hints found a bug in all my drivers using pymodbus.

Here was the old `pymodbus` code with no return type hint.
https://github.com/pymodbus-dev/pymodbus/blob/67d875b785d4f79da7b76413ae62ad71a1259fc3/pymodbus/client/serial.py#L75-L76

Here is the new `pymodbus` code with a return type hint of `None`
https://github.com/pymodbus-dev/pymodbus/blob/cfc95c4771731a0a0bc225ffcf2b4bfd1326c03f/pymodbus/transport/transport.py#L344-L345

[Here](https://github.com/numat/midas/blob/d480ec95494ae4c2993b3471049e47ec6dbb8f8f/midas/util.py#L135-L138) is my buggy code which calls the above method in pymodbus:
```python
    async def _close(self) -> None:
        """Close the TCP connection."""
        try:
            await self.client.close()  # 3.x
```

With the new type hints `mypy` reported:
`midas/util.py:138: error: "close" of "BaseTransport" does not return a value  [func-returns-value]`
The problem with the above code which had gone unnoticed until now, is that it shouldn't use `await`.

**Reply by alexrudd2** [reply]
On second thought, I think this was actually an undocumented API change?  In `3.2.0` it should be `await self.client.close()`
